Add module api for looking into INFO fields

oranagra opened the pull request:
- Add RM_GetServerInfo and friends
- Add auto memory for new opaque struct
- Add tests for new APIs

other minor fixes:
- add const in various char pointers
- requested_section in modulesCollectInfo was actually not sds but char*
- extract new string2d out of getDoubleFromObject for code reuse

Add module API for

oranagra commented:
@antirez here's my initial version of the RM_GetServerInfo.
two things that i'm considering if we should add are:

1. API for **unsigned** long long (maybe renaming `Numerical` into `LongLong` or `Integer`)
2. Improve the API by allowing the same API to look into nested dict fields, like providing a field name in the form of `db0/keys` and `cmdstat_ping/usec_per_call` (note that i did allow the RM_InfoAddField.. to create such dicts with RM_InfoBeginDictField)

MeirShpilraien commented:
@oranagra looks very good and useful.  Regarding your api suggestions:

1. Regarding unsigned long long I do not see why not add it.
2. Regarding the nested fields lookup, the current api will still hold if we want to add it in the future so we can wait to see if some module need it.

Another api I think it will be useful to have is `RM_ServerInfoGetField` that returns const char*. This api will require no extra memory allocation to return the value and will be good for someone who just need to check the value (and does not need it for future use), what do you think?

oranagra commented:
good idea @MeirShpilraien ServerInfoGetFieldC added.

regarding the unsigned, i was under the impression that we can use the same method for both (i.e. call the one that returns signed and cast the value), similarly to the fact that we only have signed function for ReplyWithLongLong, but when i tested it, i found out that string2ll errors when the value is out of range, so i renamed `Numerical` to `Signed` and added `Unsigned`.

antirez commented:
Thank you @oranagra, great work. Merged.
